MP_STATIC_ASSERT fails compiling mp_obj_is_float(mp_obj_t) as C++

I'm on esp-idf commit f333925, using xtensa-esp32-elf-gcc (crosstool-NG esp-2022r1) 11.2.0. Micropython was cloned from master about 2 weeks ago. I have the following function:

```
mp_obj_t some_func(mp_obj_t self_,mp_obj_t some_val) {
        if (mp_obj_is_float(some_val)) {
               // do something useful (no nonsense)
        }
        return mp_const_none;
    }
```

It produces the following compile error:
```
In file included from /media/EXTRA_EXT4/source_mods/micropython/py/mpstate.h:33,
                 from /media/EXTRA_EXT4/source_mods/micropython/py/runtime.h:29,
                 from /media/BACKUP/Projects/Programming/Micropython/Modules_Static/libraries/libraries.hpp:7,
                 from /media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.hpp:5,
                 from /media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp:1:
/media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp: In function 'void* some_func(mp_obj_t, mp_obj_t, mp_obj_t)':
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:54:53: error: size of array is not an integral constant-expression
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                   ~~^~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:58:40: note: in expansion of macro 'MP_STATIC_ASSERT'
   58 | #define MP_STATIC_ASSERT_NOT_MSC(cond) MP_STATIC_ASSERT(cond)
      |                                        ^~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:933:5: note: in expansion of macro 'MP_STATIC_ASSERT_NOT_MSC'
  933 |     MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:939:31: note: in expansion of macro 'mp_type_assert_not_bool_int_str_nonetype'
  939 | #define mp_obj_is_type(o, t) (mp_type_assert_not_bool_int_str_nonetype(t) && mp_obj_is_exact_type(o, t))
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:120:28: note: in expansion of macro 'mp_obj_is_type'
  120 | #define mp_obj_is_float(o) mp_obj_is_type((o), &mp_type_float)
      |                            ^~~~~~~~~~~~~~
/media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp:325:13: note: in expansion of macro 'mp_obj_is_float'
  325 |         if (mp_obj_is_float(some_val)) {
      |             ^~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:54:53: error: size of array is not an integral constant-expression
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                   ~~^~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:58:40: note: in expansion of macro 'MP_STATIC_ASSERT'
   58 | #define MP_STATIC_ASSERT_NOT_MSC(cond) MP_STATIC_ASSERT(cond)
      |                                        ^~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:934:5: note: in expansion of macro 'MP_STATIC_ASSERT_NOT_MSC'
  934 |     MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_int), assert((t) != &mp_type_int),           \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:939:31: note: in expansion of macro 'mp_type_assert_not_bool_int_str_nonetype'
  939 | #define mp_obj_is_type(o, t) (mp_type_assert_not_bool_int_str_nonetype(t) && mp_obj_is_exact_type(o, t))
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:120:28: note: in expansion of macro 'mp_obj_is_type'
  120 | #define mp_obj_is_float(o) mp_obj_is_type((o), &mp_type_float)
      |                            ^~~~~~~~~~~~~~
/media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp:325:13: note: in expansion of macro 'mp_obj_is_float'
  325 |         if (mp_obj_is_float(some_val)) {
      |             ^~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:54:53: error: size of array is not an integral constant-expression
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                   ~~^~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:58:40: note: in expansion of macro 'MP_STATIC_ASSERT'
   58 | #define MP_STATIC_ASSERT_NOT_MSC(cond) MP_STATIC_ASSERT(cond)
      |                                        ^~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:935:5: note: in expansion of macro 'MP_STATIC_ASSERT_NOT_MSC'
  935 |     MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_str), assert((t) != &mp_type_str),           \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:939:31: note: in expansion of macro 'mp_type_assert_not_bool_int_str_nonetype'
  939 | #define mp_obj_is_type(o, t) (mp_type_assert_not_bool_int_str_nonetype(t) && mp_obj_is_exact_type(o, t))
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:120:28: note: in expansion of macro 'mp_obj_is_type'
  120 | #define mp_obj_is_float(o) mp_obj_is_type((o), &mp_type_float)
      |                            ^~~~~~~~~~~~~~
/media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp:325:13: note: in expansion of macro 'mp_obj_is_float'
  325 |         if (mp_obj_is_float(some_val)) {
      |             ^~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:54:53: error: size of array is not an integral constant-expression
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                   ~~^~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/misc.h:58:40: note: in expansion of macro 'MP_STATIC_ASSERT'
   58 | #define MP_STATIC_ASSERT_NOT_MSC(cond) MP_STATIC_ASSERT(cond)
      |                                        ^~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:936:5: note: in expansion of macro 'MP_STATIC_ASSERT_NOT_MSC'
  936 |     MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_NoneType), assert((t) != &mp_type_NoneType), \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:939:31: note: in expansion of macro 'mp_type_assert_not_bool_int_str_nonetype'
  939 | #define mp_obj_is_type(o, t) (mp_type_assert_not_bool_int_str_nonetype(t) && mp_obj_is_exact_type(o, t))
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/EXTRA_EXT4/source_mods/micropython/py/obj.h:120:28: note: in expansion of macro 'mp_obj_is_type'
  120 | #define mp_obj_is_float(o) mp_obj_is_type((o), &mp_type_float)
      |                            ^~~~~~~~~~~~~~
/media/BACKUP/Projects/Programming/Micropython/Modules_Static/foo/foo.cpp:325:13: note: in expansion of macro 'mp_obj_is_float'
  325 |         if (mp_obj_is_float(some_val)) {
      |             ^~~~~~~~~~~~~~~
[1266/1285] Building CXX object esp-idf/main_esp32/CMakeFiles/__idf_main_esp32.dir/media/BACKUP/Projects/Programming/Micropython/Modules_Static/bar/bar.cpp.objninja: build stopped: subcommand failed.
ninja failed with exit code 1, output of the command is in the /tmp/build-ESP32_GENERIC-SPIRAM/log/idf_py_stderr_output_3672 and /tmp/build-ESP32_GENERIC-SPIRAM/log/idf_py_stdout_output_3672
```

--- a/examples/usercmodule/cppexample/example.cpp
+++ b/examples/usercmodule/cppexample/example.cpp
@@ -1,9 +1,16 @@
 extern "C" {
 #include <examplemodule.h>
+#include <py/objstr.h>
 
 // Here we implement the function using C++ code, but since it's
 // declaration has to be compatible with C everything goes in extern "C" scope.
 mp_obj_t cppfunc(mp_obj_t a_obj, mp_obj_t b_obj) {
+    // The following no-ops are just here to verify the static assertions used in
+    // the public API all compile with C++.
+    MP_STATIC_ASSERT_STR_ARRAY_COMPATIBLE;
+    if (mp_obj_is_type(a_obj, &mp_type_BaseException)) {
+    }
+
     // Prove we have (at least) C++11 features.
     const auto a = mp_obj_get_int(a_obj);
     const auto b = mp_obj_get_int(b_obj);

--- a/py/misc.h
+++ b/py/misc.h
@@ -52,10 +52,15 @@ typedef unsigned int uint;
 
 // Static assertion macro
 #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
-#if defined(_MSC_VER)
-#define MP_STATIC_ASSERT_NOT_MSC(cond) (1)
+// In C++ things like comparing extern const pointers are not constant-expressions so cannot be used
+// in MP_STATIC_ASSERT. Note that not all possible compiler versions will reject this. Some gcc versions
+// do, others only with -Werror=vla, msvc always does.
+// The (void) is needed to avoid "left operand of comma operator has no effect [-Werror=unused-value]"
+// when using this macro on the left-hand side of a comma.
+#if defined(_MSC_VER) || defined(__cplusplus)
+#define MP_STATIC_ASSERT_NONCONSTEXPR(cond) ((void)1)
 #else
-#define MP_STATIC_ASSERT_NOT_MSC(cond) MP_STATIC_ASSERT(cond)
+#define MP_STATIC_ASSERT_NONCONSTEXPR(cond) MP_STATIC_ASSERT(cond)
 #endif
 
 // Round-up integer division

--- a/py/obj.h
+++ b/py/obj.h
@@ -928,10 +928,10 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
 // optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type() result
 // in missed optimizations)
 #define mp_type_assert_not_bool_int_str_nonetype(t) (                                     \
-    MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
-    MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_int), assert((t) != &mp_type_int),           \
-    MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_str), assert((t) != &mp_type_str),           \
-    MP_STATIC_ASSERT_NOT_MSC((t) != &mp_type_NoneType), assert((t) != &mp_type_NoneType), \
+    MP_STATIC_ASSERT_NONCONSTEXPR((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
+    MP_STATIC_ASSERT_NONCONSTEXPR((t) != &mp_type_int), assert((t) != &mp_type_int),           \
+    MP_STATIC_ASSERT_NONCONSTEXPR((t) != &mp_type_str), assert((t) != &mp_type_str),           \
+    MP_STATIC_ASSERT_NONCONSTEXPR((t) != &mp_type_NoneType), assert((t) != &mp_type_NoneType), \
     1)
 
 #define mp_obj_is_type(o, t) (mp_type_assert_not_bool_int_str_nonetype(t) && mp_obj_is_exact_type(o, t))

